Docs/ArchitectureDecisions/004_observability.md

```diff
@@ -0,0 +1,102 @@
+
+If a user of the starter kit wants to scrape logs from modules other than LNS, or use a service other than the Application Insights SDK, the user will have to implement a custom solution.
+
+### Traces
+
+We use built-in tracing from Azure Application Insights (on an opt-in basis). This works well for function calls and correlation to other services, such as Key Vault. We will not include message flow end to end tracing for now, but will reevaluate with [#695](https://github.com/Azure/iotedge-lorawan-starterkit/issues/695).
+
+### Metrics
+
+The core modules `edgeHub` and `edgeAgent` support emitting metrics through a Prometheus endpoint, using the strategy described in [Access built-in metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-built-in-metrics?view=iotedge-2020-11). To collect these metrics and integrate everything with Application Insights, we use the metric collector (preview) as suggested in [Collect and transport metrics - Azure IoT Edge | Microsoft Docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-collect-and-transport-metrics?view=iotedge-2020-11&tabs=iothub) to export metrics to a Log Analytics storage.
+
+We will always expose LNS custom metrics in Prometheus format using [prometheus-net/prometheus-net](https://github.com/prometheus-net/prometheus-net), such that they can be consumed by any scraper that supports the Prometheus format. This will give us the following features:
+
+- Unified metrics format accross all modules in the Edge device.
+  - The Prometheus format is industrial standard understood by various consumers.
+- Decouples metrics exposure from the delivery-to-cloud approach. If at one point we decide to change how we scrap the metrics or how/where we deliver them to the observer, we can do that without changing the modules.
+- Eliminates any dependencies on Azure Monitor services (Log Analytics / Application Insights) for essential monitoring
+- Potentially gives ability to work offline if metrics are sent by the collector module through the Edge Hub using device-to-cloud channel.
+- It's up to the customer to configure how, where and what metrics to deliver from any module on an edge device.
+
+In addition to this, we will support Application Insights metrics on an opt-in basis. When enabled, we will deliver most metrics (custom and default from LNS, except the edgeAgent and edgeHub metrics, which can only be delivered to Log Analytics) to Application Insights. This will ensure that we get many of the features that we get with Application Insights out of the box (Live Metrics, integration with alerts and workbooks), while still keeping the flexibility of consuming the metrics in Prometheus format and all the advantages that come with it. This comes at the cost of increased implementation complexity.
+
```

Comment:
Perhaps it's worth mentioning here that IoT Hub comes with [curated workbooks](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-explore-curated-visualizations?view=iotedge-2020-11&tabs=devices%2Cmessaging) and a [bunch of predefined queries for alerts](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-create-alerts?view=iotedge-2020-11) based on built-in Prometheus-format metrics that are delivered by the metrics collector module to LogAnalytics.

Reply:
Addressed this with 495aabfa8cd555f40a1bf5e3607a9628a68c6f8b